### PR TITLE
[#145300245] Infrastructure for clients to request a folio swap

### PR DESCRIPTION
**Pull request by pcsouza**
@eldavido This is the proto infrastructure for implementing manual "folio swaps." A folio swap is what I'm calling the act of transferring charges between client and master folios. The reason is that it's an bidirectional term that implies the net-0 outcome of the operation. Most importantly encompasses both a swap from a client to a master folio and a swap from master to a client folio.

To do this, I'm attaching this design onto the implementation of "posted charge-credits." These are useful pseudo-transactions of a scope that matches what users expect. Rather than individual lines, a user expects to move the fundamental charge and its attendant taxes and fees.These posted charge-credits are lumped by a reference to the id of a parent transaction line. A folio swap will get included in a "posted charge-credit" by also attaching to the parent transaction line. The lines for the swaps have their own designation which will help with folio builder and client interpretation. In order to identify a "posted charge-credit" for a folio swap, I implemented an indicator for transaction lines, and included the indicator of the parent line in the charge-credit dto.

The overall goal of this service is to enable lodgeic to be naive. Lodgeic simply has to specify a posted charge-credit and the expected behavior is implicit, and unambiguous. The master folio service is responsible for ensuring the proper master-client relationship, and determining the direction that the swap occurs. By minimizing the surface area of this interface, this leaves a lot of state management business out of lodgeic and centralizes it in the money context.

A single transaction line id should be enough for the money context to construct a course of action. It can navigate from line -> account -> folio. From the folio it will be able to gather master-client relationships. Using the accounts of the master and clients, it will build a list of the history of this line's children across the master and client, and then generate a transaction attached to the line's group as a folio swap that moves the balance from one folio to the other.

### CODE REVIEW

File: proto/money/rpc/master_folio_svc.proto [hunk 1]
```diff
@@ -40,10 +41,24 @@ message MasterFolioServiceRemoveMasterFolioResponse {
 	ReservationFolioSvcRemoveMasterFolioResult result = 1;
 }
 
+enum MasterFolioSvcFolioSwapResult {
+	MASTER_FOLIO_SVC_FOLIO_SWAP_OK = 0;
+	MASTER_FOLIO_SVC_FOLIO_SWAP_IVALID_NO_MASTER_FOLIO = 1;
```
Comment (eldavido) [inline proto/money/rpc/master_folio_svc.proto]: Typo

File: proto/money/rpc/master_folio_svc.proto [hunk 2]
```diff
@@ -40,10 +41,24 @@ message MasterFolioServiceRemoveMasterFolioResponse {
...
+	MASTER_FOLIO_SVC_FOLIO_SWAP_INVALID = 2;
```
Comment (eldavido) [inline proto/money/rpc/master_folio_svc.proto]: What's the difference between this and the one above it? Is one more general?

Comment (pcsouza) [inline proto/money/rpc/master_folio_svc.proto]: I generally like having in these results a "catch-all" failure case. The above is for the specific case that there isn't an established client-master relationship for performing the swap. There are a bunch more cases though why a swap could not happen, and I didn't want to try to enumerate all of them, especially when they could be caught by the UI being a bit attentive about what requests it makes.

Comment (eldavido) [inline proto/money/rpc/master_folio_svc.proto]: Agree on the "Catch all" point. I think you should call it "unknown error" though or something a bit different. (1) Unknown or "other" makes it clear this is a catch-all, and (2) this seems too similar to the one above it.

**Comment by pcsouza** [comment]
@eldavido Changes made, PTAL

**Pull request merged**